engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbDecisionState.java

```diff
@@ -158,6 +161,28 @@ public List<PersistedDecision> findDecisionsByDecisionRequirementsKey(
     return decisions;
   }
 
+  @Override
+  public Optional<Long> findPreviousVersionDecisionKey(
+      final DirectBuffer decisionId, final int currentVersion) {
+    final Map<Integer, Long> decisionKeysByVersion = new HashMap<>();
+
+    dbDecisionId.wrapBuffer(decisionId);
+    decisionVersionByDecisionIdAndDecisionKey.whileEqualPrefix(
+        dbDecisionId,
+        ((key, version) -> {
+          if (version.getValue() < currentVersion) {
+            decisionKeysByVersion.put(version.getValue(), key.second().inner().getValue());
+          }
+        }));
```

Comment:
> Please store the entries (and iterate over them) as id -> version -> key

Good point. For some reason I was afraid this wasn't unique enough but id + version should be unique combination.

> Visiting all entries for a decision id may be expensive when many versions exist

I guess that's true. But as this is only used for deletions which is not an action that would be performed continuously I'm not too worried about this.

Reply:
>only used for deletions which is not an action that would be performed continuously

It would still block other processing, possibly leading to backpressure kicking in if this takes too long. Not sure if that's a situation we'd run into here.